api/lib/opentelemetry/context.rb

```diff
@@ -4,28 +4,126 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 
+require 'opentelemetry/context/propagation'
+
 module OpenTelemetry
-  # The Context module provides per-thread storage.
-  module Context
-    extend self
+  # Manages context on a per-thread basis
+  class Context
```

Comment:
Thinking some more about this, the [Context API](https://github.com/open-telemetry/oteps/blob/master/text/0066-separate-context-propagation.md#Context-API) from the OTEP differs in an important way from this implementation: it recommends a `CreateKey(name) -> key` function that returns a key for a name that is subsequently used to get, set and remove values.

I think the addition of this abstraction might allow a more efficient underlying implementation.

The OTEP describes two cross-cutting concerns (Observability and Correlations), while the Context API is generic enough to support additional concerns outside of the two specified. I think there is value in leveraging knowledge of the two specified concerns to improve lookup efficiency. For example, giving `Context` three fields:
```ruby
@observability = nil
@correlations = nil
@extensions = {}.freeze
```
and copying the fields in `attach` rather than creating a linked list.

My assumption here is that we can define a private class that contains all the relevant context for Observability and likewise for Correlations. For Correlations, that may well be just a frozen Hash.

Reply:
The `Context` object itself should be a general purpose context object that is not tied to observability or even OpenTelemetry. There has been some talk of extracting Context into a separate package at some point in the future. Whether or not that will happen remains to be seen, but we should design with that goal in mind.

i went ahead and reverted back to my original frozen hash design to alleviate concerns over the linked list approach. I also implemented `Context#create_key` and am using `Context::Key`s for indexing entries. I think this gives us what we need and keeps the context decoupled from observability.